datajunction-server/datajunction_server/sql/dag.py

```diff
@@ -226,6 +227,200 @@ async def get_upstream_nodes(
+    include_deactivated: bool = True,
+):
+    """
+    Get all dimension attributes for a given node.
+    """
+    node = cast(
+        Node,
+        await Node.get_by_name(
+            session,
+            node_name,
+            options=[joinedload(Node.current)],
+        ),
+    )
+    if node.type == NodeType.METRIC:
+        await refresh_if_needed(session, node.current, ["parents"])
+        node = node.current.parents[0]
+        await refresh_if_needed(session, node, ["current"])
+
+    # Discover all dimension nodes in the given node's dimensions graph
+    dimension_nodes_and_paths = await get_dimension_nodes(
+        session,
+        node,
+        include_deactivated,
+    )
+    dimensions_map = {dim.id: dim for dim, _, _ in dimension_nodes_and_paths}
+
+    # Add all reference links to the list of dimension attributes
+    reference_links = []
+    await refresh_if_needed(session, node.current, ["columns"])
+    for col in node.current.columns:
+        if col.dimension_id and col.dimension_column:
+            await session.refresh(col, ["dimension"])
+            if ref_link := await build_reference_link(  # pragma: no cover
+                session,
+                col,
+                path=[f"{node.name}.{col.name}"],
+            ):
+                reference_links.append(ref_link)
+    for dimension_node, path, role in dimension_nodes_and_paths:
+        for col in dimension_node.current.columns:
+            if col.dimension_id and col.dimension_column:
+                join_path = (
+                    [node.name] if dimension_node.name != node.name else []
+                ) + [dimensions_map[int(node_id)].name for node_id in path]
+                if ref_link := await build_reference_link(  # pragma: no cover
+                    session,
+                    col,
+                    join_path,
+                    role,
+                ):
+                    reference_links.append(ref_link)
+
+    # Build all dimension attributes from the dimension nodes in the graph
+    graph_dimensions = [
+        DimensionAttributeOutput(
+            name=f"{dim.name}.{col.name}" + (f"[{'->'.join(role)}]" if role else ""),
+            node_name=dim.name,
+            node_display_name=dim.current.display_name,
+            properties=col.attribute_names(),
+            type=str(col.type),
+            path=[node_name] + [dimensions_map[int(node_id)].name for node_id in path],
+        )
+        for dim, path, role in dimension_nodes_and_paths
+        for col in dim.current.columns
+    ]
+
+    # Build all local dimension attributes from the original node
+    local_dimensions = [
+        DimensionAttributeOutput(
+            name=f"{node.name}.{col.name}",
+            node_name=node.name,
+            node_display_name=node.current.display_name,
+            properties=col.attribute_names(),
+            type=str(col.type),
+            path=[],
+        )
+        for col in node.current.columns
```

Comment:
Should we skip some numerical columns here, or just let the user hurt themselves?

Reply:
What do you mean by numerical columns?